proposals/0057-events.md

```diff
@@ -0,0 +1,341 @@
+efficient indexing. Options must allow to paginate the result so that response
+size and latency for programs with a lot of confirmed transactions can be
+controlled. Either before or after can be used. The signature and suid are
+definining respecetively the beginning or end of the sequence which does not
+include the corresponding event. When both before and after are used the
+before option has priority, meaning the sequence starts one event before
+the specified one and ends either when reaching the limit or with the event
+after the specified one. The order of the returned event sequence is
+always in inverse chronological order, regarding how transactions are placed
+in entries and then in a block and more specifically sorted descending by the
+compound key: slot, block_entry, transaction_index_in_entry, suid.
+
+```
+EventLookupFilter: (DataSizeFilter|MemcmpFilter)
+
+rpc.getSignaturesForEvent(
+    programId: PublicKey,
+    filters?: EventLookupFilter[],
+    options?: {
+        commitment?: Commitment,
+        minContextSlot?: number,
+        limit?: number,
+        before?: {
+            signature: TransactionSignature,
+            suid: number,
+        },
+        after?: {
+            signature: TransactionSignature,
+            suid: number,
+        },
+    }
+): Promise<TransactionSignature[]>
+```
+
+In addition clients should be able to query events directly and index them
+reliably, without having to resolve signatures independently. Due to block
+re-orgs it is recommened to populate the after option with the signature and
+suid of the last known finalized event. Alternatively, the commitment level of
+the whole response can be restricted to finalized.
+
+```
+rpc.getEvents(
+    programId: PublicKey,
+    filters?: EventLookupFilter[],
+    options?: {
+        commitment?: Commitment,
+        encoding: 'base64',
+        minContextSlot?: number,
+        limit?: number,
+        before?: {
+            signature: TransactionSignature,
+            suid: number,
+        },
+        after?: {
+            signature: TransactionSignature,
+            suid: number,
+        },
+    }
+): Promise<EventResponse[]>
+
+
+EventResponse: {
+    data: Buffer,
+    instructionContext: {
+        signature: TransactionSignature,
+        suid: number,
+        outerInstructionIndex: number,
+        innerInstructionIndex?: number,
+    }
+    context: {
+        slot: number,
+        confirmationStatus: "processed" | "confirmed" | "finalized"
+    }
+}
+```
+
+For previewing transactions before signing clients should be able to receive
+events when calling `simulateTransaction`. The response should be extended
+to include an array of EventResults under the key "events".
+
+```
+EventResult: {
+    data: [string, "base64" | "base64+zstd"],
+    programId: string,
+    outerInstructionIndex:  number,
+    innerInstructionIndex?: number,
+}
+```
+
+
+### Geyser client
+
```

Comment:
we really should not be adding **new** interfaces in geyser that are also broken for non-rust languages.
we eventually need to fix the old ones, but imo we shouldn't add new ones that are also broken - please make the new `Event` structs ffi-safe.

Reply:
had a look at this PR https://github.com/anza-xyz/agave/pull/2125 and updated the events struct, lmk if that works for you